Code/GraphMol/Sgroup.h

```diff
@@ -0,0 +1,215 @@
+//
+//
+//  Copyright (C) 2002-2018 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+/*! \file Sgroup.h
+
+  \brief Defines the SGroup class
+
+*/
+#include <RDGeneral/export.h>
+#ifndef _RD_SGROUP_H
+#define _RD_SGROUP_H
+
+#include <unordered_map>
+
+#include <Geometry/point.h>
+#include <RDGeneral/types.h>
+#include <boost/smart_ptr.hpp>
+
+namespace RDKit {
+class ROMol;
+class Bond;
+class Atom;
+
+//! used to indicate errors from incorrect sgroup access
+class RDKIT_GRAPHMOL_EXPORT SGroupException : public std::runtime_error {
+ public:
+  //! construct with an error message
+  SGroupException(const char *msg) : std::runtime_error(msg){};
+  //! construct with an error message
+  SGroupException(const std::string &msg) : std::runtime_error(msg){};
+};
+
+class RDKIT_GRAPHMOL_EXPORT SGroup {
+ public:
+  enum class BondType {
+    XBOND,  // External
+    CBOND,  // Internal
+  };
+
+  typedef std::array<RDGeom::Point3D, 3> Bracket;
+
+  struct AttachPoint {
+    Atom *aAtom;
```

Comment:
I would prefer more descriptive names here.  I'm not *quite* sure what aAtom and lvAtom mean although I can guess.

Reply:
I am also not quite sure what those mean, so I just used the same names as in the SD mol specification. I figured it would be easier than to give some guessed names and then try to match them back to the specification.

Reply:
Ah, I'll take a look.  Perhaps we should have a link to the specifications somewhere near the top of the file.

Reply:
I put a link in the SGroups.md file, but I can definitely add a commet atop the header file with the link too.

Just in case you are looking for it right now: http://infochim.u-strasbg.fr/recherche/Download/Fragmentor/MDL_SDF.pdf

Reply:
The 2016 docs appear to call these:

inatom | Atom index of atom in the repeating group. | Integer > 0 |  
outatom | Atom index of atom bonded to inatom, but outside of repeating group. | Integer > 0 

which was my take on this.  Is this the same docs that you used?  Am I looking at the wrong section?

http://help.accelrysonline.com/ulm/onelab/1.0/content/ulm_pdfs/direct/reference/ctfileformats2016.pdf

Reply:
Oh! That's a more recent version than the one I have been using (the 2010 one I linked above). 

Anyway, the relevant info for the "SAP" lines (that's what the "AttachPoint" struct holds) in V3000 seems to be the same, or similar at least. My previous comment wasn't totally accurate (sorry for that): the spec mentions "aidx" and "lvidx", which I translated into the "aAtom" and "lvAtom" pointers.